### DISCUSSION TITLE
spi with software controlled slave select is behaving strange

### DISCUSSION THREAD

**Discussion by manes6969** [tabemann/zeptoforth#46] [General]
Hi ,
I was experimenting with spi , because I want to control an oled display , and decided to play a little to see how it goes. I had no trouble sending bytes to the spi. However adding a software controlled slave select behaves strange. What i do is programming a pin as output , and switching it to low , then sending the spi byte and then switching the slave select back high. This is ok when sending just 1 byte.
When sending 2 or more bytes , (in this case 3 bytes) , the slave select goes high just after the 2nd byte instead going high after the 3th byte as one would expect. I add my test program here together with 2 screeshots of my logic analyzer.
Can onyone tell me what's wrong here ? Did I forgot something , or could this be a flow in the spi of zeptoforth ...
Kind regards,
Ronny
![1byte to spi](https://github.com/tabemann/zeptoforth/assets/43147842/3a0d66fb-14bb-43d5-afcf-892c9b2d18f5)
![3bytes to spi](https://github.com/tabemann/zeptoforth/assets/43147842/f3be2a53-0b76-4561-b0ff-b30b68cbf524)
[oled.txt](https://github.com/tabemann/zeptoforth/files/11895084/oled.txt)

**Comment by tabemann** [comment]
This is probably because zeptoforth is enqueueing the bytes to send over SPI and then sending them via the SPI interrupt handler, so CS is being set high before the last byte can be dequeued and sent. A good way to avoid this is make sure to match each `>spi` call with a matching `spi>` call before you actually set CS high, to ensure that each and every byte has actually been sent first.

**Comment by tabemann** [comment]
Now that I checked, it probably is not actually enqueueing bytes in the ring buffer used by the SPI driver at this point, but rather enqueueing them in the SPI peripheral's TX FIFO, and that is not getting emptied before you set CS high. Again, make sure to use matching calls to `spi>` to ensure that every byte has actually been sent.

**Comment by manes6969** [comment]
Hi tabemann ,
Thanks for your quick reply ! Doing so makes it work .
I have 1 more question : on my logic analyzer I also notice that the timing is a bit different each time i run the code. The timing between the 3 bytes that I sent is often a bit different on each run. Is this due to the nature of this forth ? Like interrupts running all the time or something like that ?
Kind regards ,
Ronny

**Reply by tabemann** [reply]
zeptoforth normally runs with interrupts enabled, but it does disable interrupts at times. In this case either of these, ironically, can result in jitter of the timing of sent bytes. Furthermore, if you are running code that is multicore on the RP2040, each core can sometimes block the other core due to the use of hardware spinlocks and the use of a test-set mechanism based on top of hardware spinlocks (where the entire test-set mechanism shares a single hardware spinlock). All in all though, the clock signal should prevent the exact timing of bytes from being a real issue when talking to an SPI device provided the delays are not too extreme.

Travis

**Reply by manes6969** [reply]
Thanks very much for explaining.
Now I'm sure this is 'normal behaviour' .
Kind regards,
Ronny
